### PR TITLE
Implement FFI bindings

### PR DESCRIPTION
**Pull request by straight-shoota**
This patch adds a Foreign Function Interface to the compiler via bindings to [libffi](https://sourceware.org/libffi/).

We need this for interpreted mode and, similar to #11434, it imitates the behaviour of ld.so/ld for loading and calling symbols from dynamic libraries at runtime.

Also like #11434 it's a preparatory step for #11159 and adds nothing by itself. But it's an isolated non-trivial component that we can discuss it on its own.

It picks up right after the feature added in #11434, which gives you the address of a loaded symbol, but you can't really do anything with it. Through FFI you can call this function. In order to do that, libffi needs a description of the function interface, similar to a function declaration in a header file in C or a lib fun in Crystal. This description is represented as a `CallInterface`.

The major part of the implementation is already present in #11159. I just polished it up, added specs and filled out a few gaps (mostly in the C bindings).

This is just a minimal API driven by the needs of the interpreter. It's by far not a generic binding for libffi and supposed to be an internal part of the compiler.

Depends on #11434
/cc #11177

### CODE REVIEW

**Comment by straight-shoota** [comment]
The regular docker build images used for CI are missing `libffi`, hence the specs don't work there. That's why I'm temporarily switching them out for the `crystallang/crystal:ci-libffi-dev` image, which is created by a maintenance build of https://github.com/crystal-lang/distribution-scripts/pull/170

**Comment by asterite** [comment]
Totally unrelated to the contents of this PR, but if us core team members sent PR with branches in this repo, stacking PRs would be much easier to review and understand: you would be able to select a base branch, and you will be able to see a diff from just this PR, not also the PR it's based on.

Is there a reason we are not doing that? Or, put another way: is there a reason why you all are not doing that? I always do it. It's just easier. You can also use the "copy branch" link at the top to checkout the PR. Everything becomes much easier!

**Comment by j8r** [comment]
Diffing branches can always be done locally, by adding the remotes of others.

**Comment by Fryguy** [comment]
FWIW you can do a branch compare in GitHub across any forks directly in the URL (it's a little easier than trying to select it in the UI and it's more flexible).

```
https://github.com/crystal-lang/crystal/compare/FORK:REF...FORK:REF
```

where `FORK` is the name of the user (if left out it will default to the user as listed in the URL, so in the example above, `crystal-lang`), and `REF` can be any ref (e.g. branch, tag, SHA, and even can include things like `~`).

Examples (I picked a random compare branch for demonstration)
- `https://github.com/crystal-lang/crystal/compare/crystal-lang:release/1.2...straight-shoota:feature/ffi` [[link]](https://github.com/crystal-lang/crystal/compare/crystal-lang:release/1.2...straight-shoota:feature/ffi)
- `https://github.com/crystal-lang/crystal/compare/release/1.2...straight-shoota:feature/ffi` [[link]](https://github.com/crystal-lang/crystal/compare/release/1.2...straight-shoota:feature/ffi)
---

Personally, whenever I stack PRs in my branches in my day-to-day projects, I always add a helper link in the OP for reviewers that filters out the underlying PR by comparing to the underlying PR's branch ref... Something like: `https://github.com/crystal-lang/crystal/compare/Fryguy:base_pr_feature...Fryguy:my_feature`.  This way, even as I make changes including force pushes, that link will always be accurate.

**Comment by straight-shoota** [comment]
This is a draft as a preview. It's not ready for review. I'll rebase after #11434 is merged.

Please keep the discussion here on topic. We can talk about any other concerns at https://forum.crystal-lang.org/

**Review by beta-ziliani — APPROVED** [review]
👌

File: spec/compiler/data/ffi/sum.c
Comment (straight-shoota) [reply spec/compiler/data/ffi/sum.c]: Oh, maybe that's the reason why this was breaking all over CI 🤦

File: spec/compiler/data/ffi/sum.c
Comment (straight-shoota) [reply spec/compiler/data/ffi/sum.c:59]: I think I copied that from some other test. It doesn't have any significance.

File: src/compiler/crystal/loader.cr
Comment (straight-shoota) [reply src/compiler/crystal/loader.cr:1]: `ffi_spec` requires the global spec helper, which requires the entire compiler tree. That's stupid but not to resolve here.

https://github.com/crystal-lang/crystal/blob/c88be0c949143cd65eddf9a876ebb55a6ed18a55/spec/spec_helper.cr#L7

**Review by beta-ziliani — APPROVED** [review]

**Comment by straight-shoota** [comment]
I've updated the CI configuration to use the regular build images again (the 1.2.2 images have been updated to include `libffi`).

This should be ready to merge now.

**Review by asterite — APPROVED** [review]

**Comment by HertzDevil** [comment]
Don't we also need `-Dwithout_ffi` on Windows CI?

**Comment by straight-shoota** [comment]
No. The FFI bindings are currently only used from `ffi_spec`, which is skipped on non-UNIX platforms.

https://github.com/crystal-lang/crystal/blob/3b89fd689054fd8d1b4d6573690e497619c3daec/spec/compiler/ffi/ffi_spec.cr#L1

They'll actually be used with the interpreter. But even then, FFI only makes sense with dynamic loading at runtime (`Crystal::Loader`), which is not yet implemented on windows.

**Pull request merged**
